### ISSUE TITLE
Boot fails after GRUB upgrade

### ISSUE THREAD

**Issue by yabusygin** [yabusygin/extboot#13]
UEFI firmware fails to boot system after GRUB upgrade. The firmware doesn't even load GRUB.

On VirtualBox 6.1.18 the UEFI shell is invoked.

**Comment by yabusygin** [comment]
The issue is probably caused by invalid value of UEFI `Boot####` load options variable.

`Boot####` variables before GRUB upgrade:

<img width="943" alt="before" src="https://user-images.githubusercontent.com/23641097/114094695-49dafe00-98c5-11eb-8eec-344cb84cbdad.png">

`Boot####` variables after the upgrade:

<img width="943" alt="after" src="https://user-images.githubusercontent.com/23641097/114094769-5e1efb00-98c5-11eb-8f65-94e9702a9635.png">

It may be noticed that the value of `Boot0005` variable changed after the upgrade. Its value, i.e. `HD(1,MBR,0x0,0x0,0x0)`, doesn't point to external drive's ESP any more.

**Comment by yabusygin** [comment]
The bug is reproduced when GRUB is upgraded from version `2.04-1ubuntu26.8` to `2.04-1ubuntu26.9`. More pecisely, any of the following packages may be upgraded to trigger the bug:

* `grub-common`
* `grub-efi-amd64-bin` 
* `grub-efi-amd64-signed`
* `grub-pc-bin`
* `grub-pc`
* `grub2-common`

**Comment by yabusygin** [comment]
UEFI boot variables updates may be disabled by setting `grub2/update_nvram` debconf variable to `false`:

```sh
echo 'set grub2/update_nvram false' | debconf-communicate
```

See below for details.

I've examined [grub2 source package][1] (branch: `ubuntu/focal-updates`, tag: `import/2.04-1ubuntu26.9`, commit: `a614a1fdd4f02c012b5c489776dbe63e64db740d`).

`grub2/update_nvram` variable description (see `debian/templates.in`, lines 27-35):

> Update NVRAM variables to automatically boot into Debian?
>
> GRUB can configure your platform's NVRAM variables so that it boots into
> Debian automatically when powered on. However, you may prefer to disable this
> behavior and avoid changes to your boot configuration. For example, if your
> NVRAM variables have been set up such that your system contacts a PXE server
> on every boot, this would preserve that behavior.

UEFI boot variables are updated with a post install script. `grub2/update_nvram` variable value is taken into account by the script (see `debian/postinst.in`, lines 757-758). This post install script is included in `grub-pc` package.

[1]: https://launchpad.net/ubuntu/+source/grub2

**Comment by yabusygin** [comment]
> UEFI boot variables updates may be disabled by setting `grub2/update_nvram`
> debconf variable to `false`

Setting `grub2/update_nvram` to `false` doesn't fix the issue. An upgrade of [the above packages][1] still leads to modification  of the `Boot####` variable.

[1]: https://github.com/yabusygin/extboot/issues/13#issuecomment-817135597

**Comment by yabusygin** [comment]
A post-installation script of `grub-efi-amd64-signed` package ignores the value of `grub2/update_nvram` debconf variable. See below for details.

I upgraded GRUB with `/sys/firmware/efi/efivars` remounted as read-only:

```sh
mount -t efivarfs -o remount,ro,nosuid,nodev,noexec,relatime efivarfs /sys/firmware/efi/efivars
DEBIAN_FRONTEND=noninteractive apt upgrade
```

The occurred errors indicate that UEFI boot variables are updated by a post-installation script of the `grub-efi-amd64-signed` package:

![error-1](https://user-images.githubusercontent.com/23641097/114275905-f71c5600-9a2c-11eb-950c-662d55382ec8.png)

![error-2](https://user-images.githubusercontent.com/23641097/114275926-069b9f00-9a2d-11eb-9470-2f833f0690ea.png)

The `grub-efi-amd64-signed` package is built from separate `grub2-signed` source package (instead of `grub2` source package examined [above][1]).

I've explored [`grub2-signed` source package][2] (branch: `ubuntu/focal-updates`, tag: `import/1.142.11`, commit: `9426c489f0388087fcb12a557bc8602eabf322de`). The post-installation script ignores the value of `grub2/update_nvram` debconf variable and executes `grub-multi-install` with a signle argument defining a target platform (see `debian/grub-efi-amd64-signed.postinst`, line 25):

```sh
/usr/lib/grub/grub-multi-install --target=$target
```

`grub-multi-install` executes `grub-install` without adding `--no-nvram` argument:

```sh
grub-install --efi-directory=$mntpoint "$@"
```

[1]: https://github.com/yabusygin/extboot/issues/13#issuecomment-817145285
[2]: https://launchpad.net/ubuntu/+source/grub2-signed
